File: wix-petri-server/src/main/java/com/wixpress/petri/petri/PetriRpcServer.java
package com.wixpress.petri.petri;

import com.google.common.base.Predicate;
import com.google.common.base.Predicates;
import com.google.common.collect.Lists;
import com.wixpress.petri.experiments.domain.*;
import org.joda.time.DateTime;

import java.util.List;
import java.util.UUID;

import static com.google.common.base.Predicates.and;
import static com.google.common.base.Predicates.notNull;
import static com.google.common.collect.Iterables.filter;
import static com.google.common.collect.Iterables.find;
import static com.google.common.collect.Lists.newArrayList;
import static com.wixpress.petri.experiments.domain.ExperimentBuilder.aCopyOf;
import static com.wixpress.petri.experiments.domain.ExperimentPredicates.HasKey.hasKey;
import static com.wixpress.petri.experiments.domain.ExperimentPredicates.IsActivePredicate.isActiveAt;
import static com.wixpress.petri.experiments.domain.ExperimentPredicates.IsNotTerminated.isNotTerminated;
import static com.wixpress.petri.experiments.domain.ExperimentPredicates.SpecHasKey.specHasKey;
import static com.wixpress.petri.petri.PetriRpcServer.HasSpecSnapshot.hasSpecSnapshot;

/**
 * @author: talyag
 * @since: 9/9/13
 */
public class PetriRpcServer implements FullPetriClient, PetriClient , UserRequestPetriClient, PetriDeveloperApi {
    public static final String SPEC_OWNER_CHANGED_MSG = "Pay attention - Owner of %s has been changed to %s";
    public static final String SPEC_UPDATE_FAILED_MSG = "Failed to update spec [%s]";
    public static final String NON_TERMINATED_EXPERIMENTS_MSG = "Cannot update spec when non-terminated experiments exist on it";

    private final OriginalIDAwarePetriDao<Experiment, ExperimentSnapshot> experimentsDao;
    private final Clock clock;
    private final DeleteEnablingPetriDao<ExperimentSpec, ExperimentSpec> specsDao;
    private final PetriNotifier petriNotifier;
    private final MetricsReportsDao metricsReportsDao;
    private final UserStateDao userStateDao;

    public PetriRpcServer(OriginalIDAwarePetriDao<Experiment, ExperimentSnapshot> experimentsDao, Clock clock,
                          DeleteEnablingPetriDao<ExperimentSpec, ExperimentSpec> specsDao, PetriNotifier petriNotifier,
                          MetricsReportsDao metricsReportsDao, UserStateDao testGroupsDao) {
        this.experimentsDao = experimentsDao;
        this.clock = clock;
        this.specsDao = specsDao;
        this.petriNotifier = petriNotifier;
        this.metricsReportsDao = metricsReportsDao;
        this.userStateDao = testGroupsDao;
    }

    private String printOriginalAndNewSpecs(ExperimentSpec experimentSpec, ExperimentSpec originalSpec) {
        return String.format("Previous spec - [%s], new spec - [%s]", originalSpec, experimentSpec);
    }

    @Override
    public List<Experiment> fetchActiveExperiments() {
        return filterExperiments(and(notNull(), isActiveAt(clock.getCurrentDateTime())));
    }

    @Override
    public List<Experiment> fetchAllExperiments() {
        return filterExperiments(Predicates.<Experiment>notNull());
    }

    @Override
    public List<Experiment> fetchAllExperimentsGroupedByOriginalId() {
        return experimentsDao.fetchAllExperimentsGroupedByOriginalId();
    }

    private List<Experiment> filterExperiments(Predicate<Experiment> predicate) {
        List<Experiment> experiments = experimentsDao.fetch();
        Iterable<Experiment> filtered = filter(experiments, predicate);
        return newArrayList(filtered);
    }

    @Override
    public Experiment insertExperiment(ExperimentSnapshot snapshot) {
        return experimentsDao.add(snapshot);
    }

    @Override
    public Experiment updateExperiment(Experiment experiment) {
        DateTime now = clock.getCurrentDateTime();

        experimentsDao.update(experiment, now);
        return aCopyOf(experiment).withLastUpdated(now).build();
    }

    @Override
    public List<ExperimentSpec> fetchSpecs() {
        return Lists.newArrayList(filter(specsDao.fetch(), notNull()));
    }

    @Override
    public void addSpecs(List<ExperimentSpec> specs) {
        List<ExperimentSpec> existingSpecs = specsDao.fetch();
        for (ExperimentSpec spec : specs) {
            try {
                addSpec(spec, existingSpecs);
            } catch (Exception e) {
                notifyOfFailure(spec, e);
                e.printStackTrace();
            }
        }
    }

    @Override
    public List<Experiment> getHistoryById(int id) {
        return experimentsDao.getHistoryById(id);
    }

    @Override
    public void deleteSpec(String key) {
        if (!hasUnterminatedExperiments(key))
            specsDao.delete(key);
    }

    @Override
    public void reportConductExperiment(List<ConductExperimentReport> conductExperimentReports) {
        metricsReportsDao.addReports(conductExperimentReports);
    }

    @Override
    public void saveUserState(UUID userId, String userState) {
         userStateDao.saveUserState(userId, userState, new DateTime());
    }

    @Override
    public String getUserState(UUID userId) {
        return userStateDao.getUserState(userId);
    }

    @Override
    public List<ConductExperimentSummary> getExperimentReport(int experimentId) {
        return metricsReportsDao.getReport(experimentId);
    }

    private void addSpec(ExperimentSpec experimentSpec, List<ExperimentSpec> existingSpecs) {
        if (specSnapshotExists(experimentSpec, existingSpecs))
            return;

        ExperimentSpec originalSpec = findOriginalSpecByKey(existingSpecs, experimentSpec.getKey());

        if (hasUnterminatedExperiments(experimentSpec.getKey())) {
            notifyOfUpdateFailure(experimentSpec, originalSpec);
            return;
        }

        if (originalSpec == null) {
            addNew(experimentSpec);
        } else {
            updateExisting(experimentSpec, originalSpec);
        }
    }

    private void notifyOfFailure(ExperimentSpec experimentSpec, Exception e) {
        petriNotifier.notify(String.format(SPEC_UPDATE_FAILED_MSG, experimentSpec.getKey()),
                e.toString(),
                experimentSpec.getOwner());
    }

    private void notifyOfUpdateFailure(ExperimentSpec experimentSpec, ExperimentSpec originalSpec) {
        petriNotifier.notify(
                String.format(SPEC_UPDATE_FAILED_MSG, experimentSpec.getKey()) + " - " + NON_TERMINATED_EXPERIMENTS_MSG,
                printOriginalAndNewSpecs(experimentSpec, originalSpec),
                experimentSpec.getOwner());
    }

    private boolean specSnapshotExists(ExperimentSpec experimentSpec, List<ExperimentSpec> existingSpecs) {
        return find(existingSpecs, hasSpecSnapshot(experimentSpec.getExperimentSpecSnapshot()), null) != null;
    }

    private boolean hasUnterminatedExperiments(String specKey) {
        return find(fetchAllExperiments(), and(hasKey(specKey), isNotTerminated()), null) != null;
    }

    private void addNew(ExperimentSpec experimentSpec) {
        specsDao.add(experimentSpec);
    }

    private void updateExisting(ExperimentSpec experimentSpec, ExperimentSpec originalSpec) {
        DateTime now = clock.getCurrentDateTime();
        specsDao.update(experimentSpec.setCreationDate(originalSpec.getCreationDate()), now);

        notifyOwnerIfNeeded(experimentSpec, originalSpec);
    }

    private void notifyOwnerIfNeeded(ExperimentSpec experimentSpec, ExperimentSpec originalSpec) {
        if (!originalSpec.getOwner().equals(experimentSpec.getOwner())) {
            petriNotifier.notify(String.format(SPEC_OWNER_CHANGED_MSG, experimentSpec.getKey(), experimentSpec.getOwner()),
                    printOriginalAndNewSpecs(experimentSpec, originalSpec),
                    originalSpec.getOwner());
        }
    }

    private ExperimentSpec findOriginalSpecByKey(List<ExperimentSpec> existingSpecs, String key) {
        return find(existingSpecs, specHasKey(key), null);
    }

    @Override
    public UserState getFullUserState(UUID userGuid) {
        return userStateDao.getFullUserState(userGuid);
    }


    public static class HasSpecSnapshot implements Predicate<ExperimentSpec> {

        private final ExperimentSpecSnapshot snapshot;

        public HasSpecSnapshot(ExperimentSpecSnapshot snapshot) {
            this.snapshot = snapshot;
        }

        @Override
        public boolean apply(ExperimentSpec input) {
            return input.getExperimentSpecSnapshot().equals(snapshot);
        }

        public static HasSpecSnapshot hasSpecSnapshot(ExperimentSpecSnapshot snapshot) {
            return new HasSpecSnapshot(snapshot);
        }
    }

}

File: petri-server-core/src/main/java/com/wixpress/petri/petri/PetriRpcServer.scala
package com.wixpress.petri.petri

import java.util
import java.util.UUID

import com.wixpress.petri.experiments.domain.ExperimentBuilder.aCopyOf
import com.wixpress.petri.experiments.domain.{Experiment, ExperimentSnapshot, ExperimentSpec}
import com.wixpress.petri.petri.Exceptions.ExperimentNotFoundException
import com.wixpress.petri.petri.PetriRpcServer._
import org.joda.time.DateTime

import scala.collection.JavaConversions.seqAsJavaList
import scala.collection.JavaConverters._
import scala.util.control.NonFatal

/**
 * @author talyag
 * @since 9/9/13
 */
class PetriRpcServer(experimentsDao: ExperimentsDao,
                     clock: Clock,
                     specsDao: SpecsDao,
                     petriNotifier: PetriNotifier,
                     metricsReportsDao: MetricsReportsDao,
                     userStateDao: UserStateDao)
  extends FullPetriClient
  with PetriClient
  with UserRequestPetriClient
  with PetriDeveloperApi {

  override def fetchActiveExperiments(): util.List[Experiment] = {
    // check isActiveAt remains to not break not migrated stagings
    val now = clock.getCurrentDateTime
    experimentsDao.fetchBetweenStartEndDates(now).filter(exp => exp != null && exp.isActiveAt(now))
  }

  override def fetchAllExperiments(): util.List[Experiment] = {
    fetchNotNullExperiments()
  }

  override def fetchAllExperimentsGroupedByOriginalId(): util.List[Experiment] = {
    experimentsDao.fetchAllExperimentsGroupedByOriginalId
  }

  override def fetchRecentlyEndedExperimentsDueToEndDate(minutesEnded: Int): util.List[Experiment] = {
    val now = new DateTime
    val intervalStart = now.minusMinutes(minutesEnded)
    experimentsDao.fetchEndingBetween(intervalStart, now).filter(isTerminatedDueToEndDateReached)
  }

  override def fetchExperimentsGoingToEndDueToDate(minutesEnded: Int): util.List[Experiment] = {
    val now = new DateTime
    val intervalEnd = now.plusMinutes(minutesEnded)
    experimentsDao.fetchEndingBetween(now, intervalEnd)
  }

  private def fetchNotNullExperiments(): Seq[Experiment] = {
    experimentsDao.fetch().filter(exp => exp != null)
  }

  override def insertExperiment(snapshot: ExperimentSnapshot): Experiment = {
    experimentsDao.add(snapshot)
  }

  override def updateExperiment(experiment: Experiment): Experiment = {
    val now = clock.getCurrentDateTime
    experimentsDao.update(experiment, now)
    aCopyOf(experiment).withLastUpdated(now).build
  }

  override def fetchSpecs(): util.List[ExperimentSpec] = {
    specsDao.fetch().filter(_ != null)
  }

  override def addSpecs(specs: util.List[ExperimentSpec]) {
    addSpecs(specs, validate = true)
  }

  override def addSpecNoValidation(spec: ExperimentSpec) = {
    addSpecs(Seq(spec), validate = false)
  }

  override def getHistoryById(id: Int): util.List[Experiment] = {
    experimentsDao.getHistoryById(id)
  }

  override def deleteSpec(key: String) {
    if (!hasUnterminatedExperiments(key)) {
      specsDao.delete(key)
    }
  }

  override def reportConductExperiment(conductExperimentReports: util.List[ConductExperimentReport]) {
    metricsReportsDao.addReports(conductExperimentReports)
  }

  override def saveUserState(userId: UUID, userState: String) {
    userStateDao.saveUserState(userId, userState, clock.getCurrentDateTime)
  }

  override def getUserState(userId: UUID): String = {
    userStateDao.getUserState(userId)
  }

  override def getExperimentReport(experimentId: Int): util.List[ConductExperimentSummary] = {
    metricsReportsDao.getReport(experimentId)
  }

  override def getLatestExperimentReportsAfterUpdateDate(lastConductedDate: DateTime): util.List[java.lang.Integer] = {
    metricsReportsDao.getExperimentIdsLastConductedAfterGivenDate(lastConductedDate: DateTime)
  }

  private def addSpecs(specs: util.List[ExperimentSpec], validate: Boolean) {
    val existingSpecs = specsDao.fetch()
    for (spec <- specs.asScala) {
      try {
        addSpec(spec, existingSpecs, validate)
      } catch {
        case e: Exception if NonFatal(e) =>
          notifyOfFailure(spec, e)
          e.printStackTrace()
      }
    }
  }

  private def addSpec(experimentSpec: ExperimentSpec, existingSpecs: Seq[ExperimentSpec], validate: Boolean) {
    if (specSnapshotExists(experimentSpec, existingSpecs)) {
      return
    }

    val originalSpecOpt = existingSpecs.find(exp => exp.hasSameKey(experimentSpec.getKey))

    if (validate) {
      if (hasUnterminatedExperiments(experimentSpec.getKey)) {
        notifyOfUpdateFailure(experimentSpec, originalSpecOpt)
        return
      }
    }

    originalSpecOpt.fold(addNew(experimentSpec))(originalSpec => updateExisting(experimentSpec, originalSpec))
  }

  private def notifyOfFailure(experimentSpec: ExperimentSpec, e: Exception) {
    petriNotifier.notify(printSpecUpdateFailedMsg(experimentSpec.getKey), e.toString, experimentSpec.getOwner)
  }

  private def notifyOfUpdateFailure(experimentSpec: ExperimentSpec, originalSpec: Option[ExperimentSpec]) {
    petriNotifier.notify(
      printSpecUpdateFailedMsg(experimentSpec.getKey) + " - " + printNonTerminatedExperimentsMsg,
      printOriginalAndNewSpecs(experimentSpec, originalSpec),
      experimentSpec.getOwner
    )
  }

  private def specSnapshotExists(experimentSpec: ExperimentSpec, existingSpecs: Seq[ExperimentSpec]): Boolean = {
    existingSpecs.exists(es => es.getExperimentSpecSnapshot == experimentSpec.getExperimentSpecSnapshot)
  }

  private def hasUnterminatedExperiments(specKey: String): Boolean = {
    fetchNotNullExperiments().exists(exp => exp.hasSameKey(specKey) && !exp.isTerminated)
  }

  private def addNew(experimentSpec: ExperimentSpec) {
    specsDao.add(experimentSpec)
  }

  private def updateExisting(experimentSpec: ExperimentSpec, originalSpec: ExperimentSpec) {
    val now = clock.getCurrentDateTime
    specsDao.update(experimentSpec.setCreationDate(originalSpec.getCreationDate), now)
    notifyOwnerIfNeeded(experimentSpec, originalSpec)
  }

  private def notifyOwnerIfNeeded(experimentSpec: ExperimentSpec, originalSpec: ExperimentSpec) {
    if (originalSpec.getOwner != experimentSpec.getOwner) {
      petriNotifier.notify(
        printSpecOwnerChangedMsg(experimentSpec.getKey, experimentSpec.getOwner),
        printOriginalAndNewSpecs(experimentSpec, Some(originalSpec)),
        originalSpec.getOwner
      )
    }
  }

  private def isTerminatedDueToEndDateReached(experiment: Experiment): Boolean = {
    val delta = 60*1000
    Math.abs(experiment.getEndDate.getMillis - experiment.getLastUpdated.getMillis) > delta
  }

  override def getFullUserState(userGuid: UUID): UserState = {
    userStateDao.getFullUserState(userGuid)
  }

  override def migrateStartEndDates(): Unit = {
    experimentsDao.migrateStartEndDates()
  }

  @throws(classOf[ExperimentNotFoundException])
  override def fetchExperimentById(experimentId: Int): Experiment = experimentsDao.fetchExperimentById(experimentId).
    getOrElse(throw new ExperimentNotFoundException("Cannot find experiment with id = " + experimentId + "."))

}

private[petri] object PetriRpcServer {
  def printSpecOwnerChangedMsg(key: String, owner: String) = s"Pay attention - Owner of $key has been changed to $owner"

  def printSpecUpdateFailedMsg(key: String) = s"Failed to update spec [$key]"

  def printNonTerminatedExperimentsMsg = "Cannot update spec when non-terminated experiments exist on it." +
    " " +
    "If you actually want the new version you need to: terminate the existing experiment/s on this spec and then rescan your specs (via GA or manual POST request)"

  def printOriginalAndNewSpecs(experimentSpec: ExperimentSpec, originalSpec: Option[ExperimentSpec]) =
    s"Previous spec - [${originalSpec.orNull}]"+
    "  " +
    s"New spec      - [$experimentSpec]"
}
